normits_demand/models/forecasting/EDGE_growth/run.py

```diff
@@ -0,0 +1,315 @@
+            if zonal_base_demand_mx["Demand"].sum() == 0:
+                # keep matrix as it is, i.e. = 0
+                cols = zonal_base_demand_mx.columns
+                factored_matrices[segment] = ctk.pandas_utils.long_to_wide_infill(
+                    zonal_base_demand_mx,
+                    cols[0],
+                    cols[1],
+                    cols[2]
+                ).values
+                LOG.info(
+                    f"{time_period:>12}{segment:>15}" f"{0:>12}{0:>12}"
+                )
+                continue
+            # reduce probabilities to current userclass
+            hdf_file = probs_dir / f"{time_period}_iRSj_probabilities_split.h5"
+            irsj_probs_segment = pd.read_hdf(hdf_file, key=f'userclass_{userclass}')
+            # convert matrix to numpy stn2stn and produce a conversion lookup
+            if to_home:
+                zonal_base_demand_mx = utils.transpose_matrix(zonal_base_demand_mx)
+            (
+                np_stn2stn_base_demand_mx,
+                zonal_from_to_stns,
+            ) = utils.zonal_from_to_stations_demand(
+                zonal_base_demand_mx,
+                irsj_probs_segment,
+                model_stations_tlcs_len,
+                userclass,
+                time_period
+            )
+            # store matrix total demand
+            tot_input_demand = round(np_stn2stn_base_demand_mx.sum())
+            # add to overall base demand
+            overall_base_demand += tot_input_demand
+            # apply growth
+            np_stn2stn_grown_demand_mx = (
+                apply_growth.apply_demand_growth(
+                    np_stn2stn_base_demand_mx,
+                    splitting_matrices[purpose],
+                    filled_growth_matrices[purpose],
+                    len(np_stn2stn_base_demand_mx),
+                    growth_method,
+                    to_home,
+                )
+            )
+            # get movements where no growth has been applied
+            no_growth_movements = np_stn2stn_base_demand_mx[
+                np_stn2stn_base_demand_mx == np_stn2stn_grown_demand_mx
+            ]
+            # Add to the total not grown demand
+            overall_not_grown_demand = (
+                overall_not_grown_demand + no_growth_movements.sum()
+            )
+            # store matrix total demand
+            tot_output_demand = round(np_stn2stn_grown_demand_mx.sum())
+            # append to growth summary df
+            # empty dataframe for growth summary
+            segment_growth_summary = pd.DataFrame(
+                {
+                    "Time_Period": [time_period],
+                    "Demand_Segment": [segment],
+                    "Base_Demand": [tot_input_demand],
+                    f"{forecast_year}_Demand": [tot_output_demand],
+                }
+            )
+            growth_summary = pd.concat(
+                [growth_summary, segment_growth_summary], axis=0
+            )
```

Comment:
Do we need this DataFrame inside the loop at all? 

If not, its probably quicker just to store a list of dictionaries, then convert that into a DataFrame at the end of the loop.